Evolution of secondary if primary is CO + S2 and S2 flip

maxbriel opened the pull request:
In solving issue #600 with the latest v2.1-dev version, I encountered an issue with `step_detached` introduced in v2.1-dev with the separation of the matching.

When the `primary` star has become a compact object, we replace the primary's properties in `step_detached` with the secondary. This makes it possible to use the solver correctly. However, it needs to grab the values from the `secondary` star, not the `primary` to do this. In the rewrite, this got lost.

Here I reimplement it, such that the solver gives back the correct ages.
Additionally, I also solved #600 by making sure that the star initiating the RLO is always the system that becomes the main object. This stops S1 and S2 from switching.

Code cleanup + examples on the way (or available below)

sgossage commented:
@maxbriel , I was looking into this a bit more and I'm not sure if the fix we implemented fully addresses the issues. I think it's fine to accept as is, esp. if its holding up other projects, but I think there may be something up with the logic chain in `step_detached` that handles the events that may need to be addressed in the future, and I wonder if what we implemented is more of a workaround. I'll outline what I'm thinking below.

For Example 3 above, it seems like things go more or less as they should, except that `solve_ivp` terminates due to `s.t_events[3]` triggering. This event causes termination of `step_detached` if the integration time exceeds the max age of the primary. Since the primary is a more massive star, it's very short compared to the secondary. So, that triggers properly.

Also, it is correct that `secondary != binary.star_1`, since `secondary` holds the less evolved star, so `binary.event` is set to `CC1`. That's erroneous when the primary is a CO (it can't core collapse.

I think what we implemented works because we switch to using the H-rich grid in the case of a CO, which is typically of a smaller mass, and hence fairly long lived on the H-rich grid, allowing the detached step integration to proceed further and not terminate via `s.t_events[3]` .

I was checking into correcting the logic, but Quest analytics nodes have been moving at a snail's pace, so had to stop for today. In the end though, maybe switching the `htrack` is the best route.

sgossage commented:
hi @maxbriel , I think we got it right the first time after looking into it. I'd just suggest that we make the change here

```
            # states mismatch, one is a CO and other is an H or He star
            elif (all(s_valid) and (any(s_CO) and not all(s_CO))):

                ...

                # in these cases, we utilize the same htrack as the surviving secondary
                # when constructing the interp1d of the CO star
                primary.htrack = s_htrack[~CO_mask].item()
```

around line 1950 in `track_match.py`. This would set things so that whenever the primary is a CO and secondary is a normal star, we always use the `htrack` of the non-CO star.

We'd also need to remove the input of `secondary` to `get_track_match_data` if we do this.

I had a look at the logic in `step_detached`, and I do not think it is worth making changes there at this point. IMO we wouldn't have `ev_max_time2(self, t, y, primary, secondary)` as a termination event (checking if the time has exceeded that of the primary) when star is a CO. However, doing that introduces scaling errors. I think that's because the interpolators between the two stars possess very different scalings if they're very different stars.

BTW I found a slight (harmless in our cases at least) error in the wrapper for the `solve_ivp` events. It never actually sets the `terminal` property; it is hardcoded as `True`. At line 43 we have

```
def event(terminal, direction=0):
    """Return a helper function to set attributes for solve_ivp events."""
    def dec(f):
        f.terminal = True
        f.direction = direction
        return f
    return dec
```

but we should have `f.terminal = terminal` to allow someone to change things.

maxbriel commented:
I see. If we go that route, we also need to adapt the ones when there is a `massless_remnant`. 
It effectively comes down to the same solution, except where we implement it.
I don't have a strong preference. @sgossage, do you have a preferred approach?

sgossage commented:
That's right, it's the same solution, just in a different place. I agree about the massless remnant logic changing as well there. 

```
        # star 1 is a massless remnant, only star 2 exists
        elif binary.non_existent_companion == 1:
            # massless remnant
            ...
            primary.htrack = s_htrack[1]
           ...

        # star 2 is a massless remnant, only star 1 exists
        elif binary.non_existent_companion == 2:
            ...
            primary.htrack = s_htrack[0]
            ...
```

I prefer making the changes in `determine_star_states` if we can so that the logic handling these changes is all contained in one spot. Otherwise we have assignments happening for `htrack` there, and then again conditionally in `get_star_match_data`.

maxbriel commented:
@sgossage I've changed it to do it in `determine_star_state`. 

While testing with a population, I ran into more issues, which was why I made the `mb_matching_fix` branch and PR #657. I merged that PR into this branch, such that it's easier to keep track of everything.
It adds the following:

- Fixes the output structure for "later" matching attempts, such that the right information is passed along, if a matched failed.
- Renames a variable such that it's correctly used in the output. Otherwise, the variable is undefined.
- Fixes that interp1d might not be set for the primary or secondary, instead we check for the existence of the variable.

It mostly housekeeping of "failed" outputs to make sure the structure is similar to what's expected.

I also found that the `accreted_He_Core_He_burning` state was originally matched to Helium star, but failed here with an invalid state (see Issue #659 ). I added it back to the He matching list.

maxbriel commented:
It looks like the result/track is already super close to the final value. I added a "tol=1e-6" to the scipy minimizer, which allows some of them to pass correctly with the Powell method (which was hitting maxfunc calls).
They still fail in the initial TNC minimisation due to the initial linear search failing.

So this is not really a good solution.

sgossage commented:
hey @maxbriel , can set if setting unbounded upper age limits for the minimization routine resolves the remaining issues? Lines 350 and 352 may need to change as below to achieve the same behavior as v2.0.0.

```
348        # min/max ranges of ages for each grid
349        t_min_H = 0.0
350        t_max_H = None
351        t_min_He = 0.0
352        t_max_He = None
```

sgossage commented on posydon/binary_evol/DT/track_match.py:
```diff
@@ -1100,7 +1101,7 @@ def get_match_attrs(match_type="default", match_ok=True):
                                 "since star is on the MS. Skipping...")
                     match_ok = False
-                    return None, None, False, match_ok
+                    return (None, None, False, match_ok), (None, None, None)
```
```suggestion
                    match_attrs = (None, None, new_htrack, match_ok)
                    scls_bnds = (None, None, None)
                    return match_attrs, scls_bnds
```
